src/Pos/Update/Poll/Pure.hs

```diff
@@ -0,0 +1,88 @@
+-- | Pure Poll
+
+module Pos.Update.Poll.Pure
+       ( PurePoll (..)
+       ) where
+
+import           Universum
+
+import           Control.Lens               (at, to, (%=), (.=))
+import qualified Data.HashMap.Strict        as HM
+import           System.Wlog                (CanLog, HasLoggerName, NamedPureLogger,
+                                             logWarning)
+
+import           Pos.Crypto                 (hash)
+import           Pos.Types                  (SoftwareVersion (..))
+import           Pos.Update.Poll.Class      (MonadPoll (..), MonadPollRead (..))
+import qualified Pos.Update.Poll.PollState  as Poll
+import           Pos.Update.Poll.Types      (BlockVersionState (..),
+                                             DecidedProposalState (..),
+                                             UndecidedProposalState (..),
+                                             cpsSoftwareVersion, propStateToEither)
+
+newtype PurePoll a = PurePoll
+    { getPurePoll :: NamedPureLogger (State Poll.PollState) a
+    } deriving (Functor, Applicative, Monad, CanLog, HasLoggerName)
+
+instance MonadPollRead PurePoll where
+    getBVState bv = PurePoll $ use $ Poll.psBlockVersions . at bv
+    getProposedBVs = PurePoll $ use $ Poll.psBlockVersions . to HM.keys
+    getEpochProposers = PurePoll $ use $ Poll.psEpochProposers
+    getConfirmedBVStates = PurePoll $ use $ Poll.psBlockVersions . to HM.toList
+    getAdoptedBVFull = PurePoll $ use $ Poll.psAdoptedBV
+    getLastConfirmedSV an = PurePoll $ use $ Poll.psConfirmedBVs . at an
+    getProposal ui = PurePoll $ use $ Poll.psActiveProposals . at ui
+    getProposalsByApp an = PurePoll $ do
+        delActiveProposalsIndices <- use Poll.psDelActivePropsIdx
+        activeProposals           <- use Poll.psActiveProposals
+        pure $ propGetByApp delActiveProposalsIndices activeProposals
+      where propGetByApp appHashmap upIdHashmap = fromMaybe [] $
+                do hashset <- HM.lookup an appHashmap
+                   let uidList = toList hashset
+                       propStateList = fmap (flip HM.lookup upIdHashmap) uidList
+                   sequence propStateList
+    getConfirmedProposals = PurePoll $ use $ Poll.psConfirmedProposals . to HM.elems
+    getEpochTotalStake ei =
+        PurePoll $ use $ Poll.psFullRichmenData . to ((Just . fst) <=< HM.lookup ei)
+    getRichmanStake ei si =
+        PurePoll $ use $ Poll.psFullRichmenData .
+            to ((HM.lookup si . snd) <=< HM.lookup ei)
+    getOldProposals si =
+        PurePoll $ use $ Poll.psActiveProposals . to (filter ((<= si) . upsSlot) .
+                                                      lefts .
+                                                      map propStateToEither .
+                                                      HM.elems)
+    getDeepProposals cd =
+        PurePoll $ use $ Poll.psActiveProposals .
+            to (filter ((> (Just cd)) . dpsDifficulty) .
```

Comment:
Wait, why is predicate `> (Just cd)`? Shouldn't it be `<=`? Also how will it work if `dpsDifficulty` is `Nothing`? If it's `Nothing`, its depth is basically `0`, so it should be always `False`.